File: java/Tutorials/BiometricStandards/C/ANTemplateType17FromNImage/ANTemplateType17FromNImage.c
#include <TutorialUtils.h>

#ifdef N_MAC_OSX_FRAMEWORKS
	#include <NCore/NCore.h>
	#include <NBiometricClient/NBiometricClient.h>
	#include <NBiometrics/NBiometrics.h>
	#include <NMedia/NMedia.h>
	#include <NLicensing/NLicensing.h>
#else
	#include <NCore.h>
	#include <NBiometricClient.h>
	#include <NBiometrics.h>
	#include <NMedia.h>
	#include <NLicensing.h>
#endif

const NChar title[] = N_T("ANTemplateType17FromNImage");
const NChar description[] = N_T("Demonstrates creation of ANTemplate with type 17 record in it");
const NChar version[] = N_T("9.0.0.0");
const NChar copyright[] = N_T("Copyright (C) 2009-2017 Neurotechnology");

int usage()
{
	printf(N_T("usage:\n"));
	printf(N_T("\t%s [NImage] [ANTemplate] [Tot] [Dai] [Ori] [Tcn] [Src]\n"), title);
	printf(N_T("\n"));
	printf(N_T("\tNImage     - filename with image file\n"));
	printf(N_T("\tANTemplate - filename for ANTemplate\n"));
	printf(N_T("\tTot - specifies type of transaction\n"));
	printf(N_T("\tDai - specifies destination agency identifier\n"));
	printf(N_T("\tOri - specifies originating agency identifier\n"));
	printf(N_T("\tTcn - specifies transaction control number\n"));
	printf(N_T("\tSrc - specifies source agency number\n"));

	return 1;
}

NResult ANTemplateAddRecordType17(
	NChar * szFileNameIn, // pointer to string that specifies filename of image file
	NChar * szFileNameOut, // pointer to string that specifies filename to store ANTemplate
	const NChar * szTot, // pointer to tring that specifies type of transaction
	const NChar * szDai, // pointer to string that specifies destination agency identifier
	const NChar * szOri, // pointer to string that specifies originating agency identifier
	const NChar * szTcn, // pointer to string that specifies transaction control number
	const NChar * szSrc	// pointer to string that specifies source agency number
	)
{
	HANTemplate hANTemplate = NULL; // handle to ANTemplate object
	HNImage hImage = NULL; // handle to NImage object
	HNImage hRGBImage = NULL; // handle to RGB object
	HANType17Record hRecord = NULL; // handle to ANRecord object
	NResult result;
	NInt idc = 0;
	ANImageCompressionAlgorithm cga = anicaNone; // image compression algortihm currently none
	BdifScaleUnits slc = bsuNone;

	// Create empty ANTemplate object with only type 1 record in it
	result = ANTemplateCreateWithTransactionInformation(AN_TEMPLATE_VERSION_CURRENT, szTot, szDai, szOri, szTcn, 0, &hANTemplate);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("error in ANTemplateCreateWithTransactionInformation, error code: %d"), result);
		goto FINALLY;
	}

	// Create NImage object from image file
	result = NImageCreateFromFileEx(szFileNameIn, NULL, 0, NULL, &hImage);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("error in NImageCreateFromFile, error code: %d"), result);
		goto FINALLY;
	}

	result = NImageSetResolutionIsAspectRatio(hImage, NTrue);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("error in NImageSetResolutionIsAspectRatio, error code: %d"), result);
		goto FINALLY;
	}

	// create RGB image from an image
	result = NImagesRecolorImage(hImage, NPF_RGB_8U, NULL, NULL, &hRGBImage);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("error in NImagesRecolorImage, error code: %d"), result);
		goto FINALLY;
	}

	// Create Type 17 record
	result = ANType17RecordCreateFromNImage(AN_TEMPLATE_VERSION_CURRENT, idc, szSrc, slc, cga, hRGBImage, 0, &hRecord);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("error in ANType17RecordCreateFromNImage, error code: %d"), result);
		goto FINALLY;
	}

	// Add Type 17 record to ANTemplate object
	result = ANTemplateAddRecordEx(hANTemplate, hRecord, NULL);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("error in ANTemplateAddRecordEx, error code: %d"), result);
		goto FINALLY;
	}

	// Store ANTemplate object with type 17 record in file
	result = ANTemplateSaveToFile(hANTemplate, szFileNameOut, 0);
	if(NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("error in ANTemplateSaveToFile, error code: %d"), result);
		goto FINALLY;
	}

	printf(N_T("template saved to %s"), szFileNameOut);

	result = N_OK;
FINALLY:
	{
		NResult result2 = NObjectSet(NULL, &hRecord);
		if NFailed((result2)) PrintErrorMsg(N_T("NObjectSet() failed (result = %d)!"), result2);
		result2 = NObjectSet(NULL, &hANTemplate);
		if NFailed((result2)) PrintErrorMsg(N_T("NObjectSet() failed (result = %d)!"), result2);
		result2 = NObjectSet(NULL, &hImage);
		if NFailed((result2)) PrintErrorMsg(N_T("NObjectSet() failed (result = %d)!"), result2);
		result2 = NObjectSet(NULL, &hRGBImage);
		if NFailed((result2)) PrintErrorMsg(N_T("NObjectSet() failed (result = %d)!"), result2);
	}

	return result;
}

int main(int argc, NChar **argv)
{
	NResult result = N_OK;
	const NChar * components = N_T("Biometrics.Standards.Irises");
	NBool available = NFalse;

	OnStart(title, description, version, copyright, argc, argv);

	if(argc < 8)
	{
		OnExit();
		return usage();
	}

	// check the license first
	result = NLicenseObtainComponents(N_T("/local"), N_T("5000"), components, &available);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("NLicenseObtainComponents() failed, result = %d\n"), result);
		goto FINALLY;
	}

	if (!available)
	{
		printf(N_T("Licenses for %s not available\n"), components);
		result = N_E_NOT_ACTIVATED;
		goto FINALLY;
	}

	result = ANTemplateAddRecordType17(argv[1], argv[2], argv[3], argv[4], argv[5], argv[6], argv[7]);
	if (NFailed(result))
	{
		goto FINALLY;
	}

	result = N_OK;
FINALLY:
	{
		NResult result2 = NLicenseReleaseComponents(components);
		if (NFailed(result2)) PrintErrorMsg(N_T("NLicenseReleaseComponents() failed, result = %d\n"), result2);
	}

	OnExit();
	return result;
}

File: java/Tutorials/BiometricStandards/Java/antemplate-type17-from-nimage/src/main/java/com/neurotec/tutorials/ANTemplateType17FromNImage.java
package com.neurotec.tutorials;

import java.io.IOException;

import com.neurotec.biometrics.standards.ANImageCompressionAlgorithm;
import com.neurotec.biometrics.standards.ANTemplate;
import com.neurotec.biometrics.standards.ANType17Record;
import com.neurotec.biometrics.standards.BDIFScaleUnits;
import com.neurotec.images.NImage;
import com.neurotec.images.NPixelFormat;
import com.neurotec.licensing.NLicense;
import com.neurotec.tutorials.util.LibraryManager;
import com.neurotec.tutorials.util.Utils;

public final class ANTemplateType17FromNImage {

	private static final String DESCRIPTION = "Demonstrates creation of ANTemplate with type 17 record in it";
	private static final String NAME = "antemplate-type17-from-nimage";
	private static final String VERSION = "9.0.0.0";

	private static void usage() {
		System.out.println("usage:");
		System.out.format("\t%s [NImage] [ANTemplate] [Tot] [Dai] [Ori] [Tcn] [Src]%n", NAME);
		System.out.println("");
		System.out.println("\t[NImage]     - filename with image file.");
		System.out.println("\t[ANTemplate] - filename for ANTemplate.");
		System.out.println("\t[Tot] - specifies type of transaction.");
		System.out.println("\t[Dai] - specifies destination agency identifier.");
		System.out.println("\t[Ori] - specifies originating agency identifier.");
		System.out.println("\t[Tcn] - specifies transaction control number.");
		System.out.println("\t[Src] - specifies source agency number.");
		System.out.println("");
	}

	public static void main(String[] args) {
		LibraryManager.initLibraryPath();
		final String components = "Biometrics.Standards.Irises";

		Utils.printTutorialHeader(DESCRIPTION, NAME, VERSION, args);

		if (args.length != 7) {
			usage();
			System.exit(1);
		}

		ANTemplate template = null;
		NImage image = null;
		NImage rgbImage = null;

		try {
			if (!NLicense.obtainComponents("/local", 5000, components)) {
				System.err.println("Could not obtain licenses for components: " + components);
				System.exit(-1);
			}

			String tot = args[2]; // type of transaction
			String dai = args[3]; // destination agency identifier
			String ori = args[4]; // originating agency identifier
			String tcn = args[5]; // transaction control number
			String src = args[6]; // source agency number

			if ((tot.length() < 3) || (tot.length() > 4)) {
				System.out.println("Tot parameter should be 3 or 4 characters length.");
				System.exit(-1);
			}

			// Create empty ANTemplate object with only type 1 record in it
			template = new ANTemplate(ANTemplate.VERSION_CURRENT, tot, dai, ori, tcn, 0);

			// Open image
			image = NImage.fromFile(args[0]);

			// Convert to rgb image
			rgbImage = NImage.fromImage(NPixelFormat.RGB_8U, 0, image);

			rgbImage.setResolutionIsAspectRatio(true);

			// Add Type 17 record to ANTemplate object
			ANType17Record record = new ANType17Record(ANTemplate.VERSION_CURRENT, 0, src, BDIFScaleUnits.NONE, ANImageCompressionAlgorithm.NONE, rgbImage);
			template.getRecords().add(record);

			// Store ANTemplate object with type 17 record in file
			template.save(args[1]);
		} catch (Throwable th) {
			Utils.handleError(th);
		} finally {
			try {
				NLicense.releaseComponents(components);
			} catch (IOException e) {
				e.printStackTrace();
			}
			if (template != null) template.dispose();
			if (image != null) image.dispose();
			if (rgbImage != null) rgbImage.dispose();
		}
	}
}
